cps/kobo.py

```diff
@@ -0,0 +1,544 @@
+            "BookMetadata": get_metadata(book),
+            "ReadingState": reading_state(book),
+        }
+        if book.timestamp > sync_token.books_last_created:
+            entitlements.append({"NewEntitlement": entitlement})
+        else:
+            entitlements.append({"ChangedEntitlement": entitlement})
+
+        new_books_last_modified = max(
+            book.last_modified, sync_token.books_last_modified
+        )
+        new_books_last_created = max(book.timestamp, sync_token.books_last_modified)
+
+    sync_token.books_last_created = new_books_last_created
+    sync_token.books_last_modified = new_books_last_modified
+
+    # Missing feature: Detect server-side book deletions.
+
+    # Missing feature: Join the response with results from the official Kobo store so that users can still buy and access books from the device store (particularly while on-the-road).
+
+    response = make_response(jsonify(entitlements))
+
+    sync_token.to_headers(response.headers)
+    response.headers["x-kobo-sync-mode"] = "delta"
+    response.headers["x-kobo-apitoken"] = "e30="
+    return response
+
+
+@kobo.route("/v1/library/<book_uuid>/metadata")
+@login_required
+@download_required
+def HandleMetadataRequest(book_uuid):
+    log.info("Kobo library metadata request received for book %s" % book_uuid)
+    book = db.session.query(db.Books).filter(db.Books.uuid == book_uuid).first()
+    if not book or not book.data:
+        log.info(u"Book %s not found in database", book_uuid)
+        return make_response("Book not found in database.", 404)
+
+    metadata = get_metadata(book)
+    return jsonify([metadata])
+
+
+def get_download_url_for_book(book, book_format):
+    return "{url_base}/download/{book_id}/{book_format}".format(
+        url_base=config.config_server_url,
+        book_id=book.id,
+        book_format=book_format.lower(),
+    )
+
+
+def get_download_url_for_book_b2(book, book_format):
+    if not data:
+        log.info(u"Book %s does have a kepub format", book_uuid)
+        return None
+
+    file_name = data.name + ".kepub"
+    file_path = os.path.join(book.path, file_name)
+
+    if not os.path.isfile(B2_SECRETS):
+        log.error(u"b2 secret file not found")
+        return None
+    with open(B2_SECRETS, "r") as filedata:
+        secrets = json.load(filedata)
+
+    info = InMemoryAccountInfo()
+    b2_api = B2Api(info)
+    b2_api.authorize_account(
+        "production", secrets["application_key_id"], secrets["application_key"]
+    )
+    bucket = b2_api.get_bucket_by_name(secrets["bucket_name"])
+    if not bucket:
+        log.error(u"b2 bucket not found")
+        return None
+
+    download_url = b2_api.get_download_url_for_file_name(
+        secrets["bucket_name"], file_path
+    )
+    download_authorization = bucket.get_download_authorization(
+        file_path, valid_duration_in_seconds=600
+    )
+    return download_url + "?Authorization=" + download_authorization
```

Comment:
This function can be removed? Only needed for the backblaze integration

Reply:
Done.